File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/f18bb803b8807aa5e0dedae871ef4247.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

type B1T struct {
	bit []int
	n   int
}

func (b *B1T) add(djg, x int) {
	for djg <= b.n {
		b.bit[djg] += x
		djg += (djg & -djg)
	}
}

func (b *B1T) sum(djg int) int {
	s0m := 0
	for djg > 0 {
		s0m += b.bit[djg]
		djg -= (djg & -djg)
	}
	return s0m
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	n := nextInt(scanner)
	w := nextInt(scanner)

	x := make([]float64, n)
	y := make([]float64, n)
	cnt := make(map[float64]int)

	for i := 0; i < n; i++ {
		a := nextInt(scanner)
		b := nextInt(scanner)
		x[i] = float64(a) / float64(b+w)
		y[i] = float64(a) / float64(b-w)
		cnt[y[i]]++
	}

	ans := 0
	for i := 0; i < n; i++ {
		ans += cnt[y[i]] - 1
	}
	ans /= 2

	as := make([]float64, n)
	copy(as, x)
	sort.Float64s(as)
	cnta := 0
	for i := 0; i < n; i++ {
		if i == 0 || as[i] != as[i-1] {
			as[cnta] = as[i]
			cnta++
		}
	}

	bs := make([]pair, n)
	for i := 0; i < n; i++ {
		bs[i] = pair{y[i], i}
	}
	sort.Slice(bs, func(i, j int) bool {
		return bs[i].first < bs[j].first
	})

	bit := B1T{bit: make([]int, 100007), n: 100007}
	now := 0
	for i := 0; i < n; i++ {
		for bs[now].first < bs[i].first {
			id := sort.SearchFloat64s(as, x[bs[now].second])
			bit.add(100006-id, 1)
			now++
		}
		id := sort.SearchFloat64s(as, x[bs[i].second])
		ans += bit.sum(100006 - id)
	}

	fmt.Println(ans)
}

type pair struct {
	first  float64
	second int
}

func nextInt(scanner *bufio.Scanner) int {
	scanner.Scan()
	num, _ := strconv.Atoi(scanner.Text())
	return num
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/f18bb803b8807aa5e0dedae871ef4247.java
import java.util.*;

class Main {
    static int n;
    static double[] x, y;
    static int w;
    static Map<Double, Integer> cnt;
    static double[] as;
    static int cnta;
    static Pair[] bs;
    static BIT bit;

    static class BIT {
        static final int N = 100007;
        long[] bit = new long[N];

        void add(int djg, int x) {
            while (djg <= N) {
                bit[djg] += x;
                djg += (djg & -djg);
            }
        }

        long sum(int djg) {
            long s0m = 0;
            while (djg > 0) {
                s0m += bit[djg];
                djg -= (djg & -djg);
            }
            return s0m;
        }
    }

    static class Pair implements Comparable<Pair> {
        double first;
        int second;

        Pair(double first, int second) {
            this.first = first;
            this.second = second;
        }

        public int compareTo(Pair p) {
            if (this.first < p.first) {
                return -1;
            } else if (this.first > p.first) {
                return 1;
            } else {
                return 0;
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        w = sc.nextInt();
        x = new double[n];
        y = new double[n];
        cnt = new HashMap<>();
        as = new double[n];
        bs = new Pair[n];
        bit = new BIT();

        for (int i = 0; i < n; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            x[i] = 1.00 * a / (b + w);
            y[i] = 1.00 * a / (b - w);
            cnt.put(y[i], cnt.getOrDefault(y[i], 0) + 1);
        }

        long ans = 0;
        for (int i = 0; i < n; i++) {
            ans += cnt.get(y[i]) - 1;
        }
        ans /= 2;

        for (int i = 0; i < n; i++) {
            as[i] = x[i];
        }
        Arrays.sort(as);
        cnta = 0;
        for (int i = 0; i < n; i++) {
            if (i == 0 || as[i] != as[i - 1]) {
                as[cnta++] = as[i];
            }
        }

        for (int i = 0; i < n; i++) {
            bs[i] = new Pair(y[i], i);
        }
        Arrays.sort(bs);

        int now = 0;
        for (int i = 0; i < n; i++) {
            while (bs[now].first < bs[i].first) {
                int id = Arrays.binarySearch(as, 0, cnta, x[bs[now].second]);
                bit.add(100006 - id, 1);
                now++;
            }
            int id = Arrays.binarySearch(as, 0, cnta, x[bs[i].second]);
            ans += bit.sum(100006 - id);
        }

        System.out.println(ans);
    }
}
